presto-native-execution/presto_cpp/main/expression/RowExpressionEvaluator.cpp

```diff
@@ -0,0 +1,419 @@
+    } else {
+      // call
+      res["@type"] = "call";
+      protocol::Signature signature;
+      std::string exprName = expr->name();
+      if (scalarOperatorMap_.find(expr->name()) != scalarOperatorMap_.end()) {
+        exprName = scalarOperatorMap_.at(expr->name());
+      }
+      signature.name = exprName;
+      res["displayName"] = exprName;
+      signature.kind = protocol::FunctionKind::SCALAR;
+      signature.typeVariableConstraints = {};
+      signature.longVariableConstraints = {};
+      signature.returnType = expr->type()->toString();
+
+      std::vector<protocol::TypeSignature> argumentTypes;
+      auto inputs = expr->inputs();
+      auto numArgs = inputs.size();
+      argumentTypes.reserve(numArgs);
+      for (auto i = 0; i < numArgs; i++) {
+        argumentTypes.emplace_back(inputs[i]->type()->toString());
+      }
+      signature.argumentTypes = argumentTypes;
+      signature.variableArity = false;
+
+      protocol::BuiltInFunctionHandle builtInFunctionHandle;
+      builtInFunctionHandle._type = "$static";
+      builtInFunctionHandle.signature = signature;
+      res["functionHandle"] = builtInFunctionHandle;
+      res["returnType"] = expr->type()->toString();
+      for (const auto& input : inputs) {
+        res["arguments"].push_back(veloxExprToRowExpr(input, inputRowExpr));
+      }
+    }
+  } else {
+    VELOX_USER_FAIL(
+        "Unable to convert Velox Expr to Presto RowExpression: {}",
+        expr->toString());
+  }
+
+  return res;
+}
+
+// Compiles protocol::RowExpression into a velox expression, with constant
+// folding enabled.
+std::shared_ptr<const velox::exec::Expr>
+RowExpressionEvaluator::compileExpression(
+    const std::shared_ptr<protocol::RowExpression>& inputRowExpr) {
+  auto typedExpr = exprConverter_.toVeloxExpr(inputRowExpr);
+  exec::ExprSet exprSet{{typedExpr}, execCtx_.get()};
+  auto compiledExprs =
+      exec::compileExpressions({typedExpr}, execCtx_.get(), &exprSet, true);
+  return compiledExprs[0];
+}
+
+void RowExpressionEvaluator::evaluateExpression(
+    const std::vector<std::unique_ptr<folly::IOBuf>>& body,
+    proxygen::ResponseHandler* downstream) {
+  try {
+    std::ostringstream oss;
+    for (auto& buf : body) {
+      oss << std::string((const char*)buf->data(), buf->length());
+    }
+    json::array_t inputList = json::parse(oss.str());
+    auto numExpr = inputList.size();
+    json output = json::array();
+
+    for (auto i = 0; i < numExpr; i++) {
+      std::shared_ptr<protocol::RowExpression> inputRowExpr = inputList[i];
+      if (const auto special =
+              std::dynamic_pointer_cast<protocol::SpecialFormExpression>(
+                  inputRowExpr)) {
+        inputRowExpr = optimizeSpecialForm(special);
+      }
+      const auto compiledExpr = compileExpression(inputRowExpr);
+      json resultJson = veloxExprToRowExpr(compiledExpr, inputList[i]);
+      output.push_back(resultJson);
+    }
+
+    proxygen::ResponseBuilder(downstream)
```

Comment:
This response building shouldn't be part of this class and function.

Reply:
Moved response building to function `evaluate`, and the expression evaluation to `evaluateExpression` as suggested.